File: 프로그래머스/3/67258. ［카카오 인턴］ 보석 쇼핑/［카카오 인턴］ 보석 쇼핑.py
def solution(gems):
    dic = {gems[0]:1}
    s = 0
    e = 0
    answer = [0,len(gems)-1]
    total_num = len(set(gems))
    while s<=e<len(gems):
        if len(dic) < total_num:
            e += 1
            if e == len(gems):
                break
            dic[gems[e]] = dic.get(gems[e],0)+1
        else:
            if (e-s+1) < answer[1]-answer[0]+1:
                answer[0],answer[1]= s, e
            if dic.get(gems[s])==1:
                del dic[gems[s]]
            else:
                dic[gems[s]] -= 1
            s += 1
    answer[0] +=1
    answer[1] +=1
    return answer

File: 프로그래머스/3/67258. ［카카오 인턴］ 보석 쇼핑/［카카오 인턴］ 보석 쇼핑.java
import java.util.*;
class Solution {
    public int[] solution(String[] gems) {
        int[] answer = {};
        int ans_left = 0;
        int ans_right = 0;
        int left = 0;
        Map<String,Integer> map = new HashMap<>();
        Set<String> set = new HashSet<>();
        int distance = Integer.MAX_VALUE;
        for(int i=0;i<gems.length;i++){
            set.add(gems[i]);
        }
        for(int right=0;right<gems.length;right++){
            map.put(gems[right],map.getOrDefault(gems[right],0)+1);
            while(map.get(gems[left])>1){
                map.put(gems[left],map.get(gems[left])-1);
                left += 1;
            }
            if(map.size()==set.size() && distance>right-left){
                distance = right-left;
                ans_left = left+1;
                ans_right = right+1;
            }
        }
        return new int[]{ans_left, ans_right};
    }
}